projects/concurrency/main.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"fmt"
+	"time"
+)
+
+type Cache struct {
+	capacity int
+	storage  map[int]entry
+}
+
+type entry struct {
+	value    any
+	lastUsed time.Time
+}
+
+func NewCache(capacity int) *Cache {
+	return &Cache{
+		capacity: capacity,
+		storage:  make(map[int]entry, capacity),
+	}
+}
+
+// Put adds the value to the cache, and returns a boolean to indicate whether a value already existed in the cache for that key.
+// If there was previously a value, it replaces that value with this one.
+// Any Put counts as a refresh in terms of LRU tracking.
+func (c *Cache) Put(key int, value any) bool {
+	// Check if the entry already exists in the cache
+	_, exists := c.storage[key]
+	// Update the entry with the new value and current time
+	c.storage[key] = entry{value: value, lastUsed: time.Now()}
+	// If the cache is at capacity, remove the least recently used (LRU) entry
+	if len(c.storage) > c.capacity {
+		var lruKey int
+		var lruTime time.Time
+		for k, e := range c.storage {
```

Comment:
This works, but it requires looking through every element in the map in order to remove one (i.e. it is `O(n)`). Let's imagine you had a million elements in your cache, this means once it becomes full, every time you put one value in, you need to look at 1M elements. So if you put 1M new elements in, you would need to look at 1 _trillion_ elements.

Can you think of a way to not need to look through the entire map each time you need to evict something?

(This is less important to focus on than the concurrency aspects of this project, so I'd probably come back to this after adding the currency pieces)

Reply:
I knew that I need to use linked list , but since I have a limited time , I wanted to do the simplest thing and focus on the concurrency aspects :)